src/workerd/jsg/filesystem.c++

```diff
@@ -0,0 +1,537 @@
+    if (size == owned.size()) return;  // Nothing to do.
+
+    auto newData = kj::heapArray<kj::byte>(size);
+
+    if (size > owned.size()) {
+      // To grow the file, we need to allocate a new array, copy the old data over,
+      // and replace the original.
+      newData.first(owned.size()).copyFrom(owned);
+      newData.slice(owned.size()).fill(0);
+    } else {
+      newData.asPtr().copyFrom(owned.first(size));
+    }
+    ownedOrView = newData.attach(js.getExternalMemoryAdjustment(newData.size()));
+  }
+
+  void zero(Lock& js, kj::Maybe<size_t> offset) override {
+    auto& owned = JSG_REQUIRE_NONNULL(
+        ownedOrView.tryGet<kj::Array<kj::byte>>(), Error, "Cannot zero a read-only file");
+    owned.slice(offset.orDefault(0)).fill(0);
+  }
+
+  kj::StringPtr jsgGetMemoryName() const override {
+    return "File"_kj;
+  }
+
+  size_t jsgGetMemorySelfSize() const override {
+    return sizeof(FileImpl);
+  }
+
+  void jsgGetMemoryInfo(jsg::MemoryTracker& tracker) const override {
+    // We only track the memory if we own the data.
+    KJ_SWITCH_ONEOF(ownedOrView) {
+      KJ_CASE_ONEOF(owned, kj::Array<kj::byte>) {
+        tracker.trackField("owned", owned);
+        return;
+      }
+      KJ_CASE_ONEOF(view, kj::ArrayPtr<const kj::byte>) {
+        return;
+      }
+    }
+  }
+
+ private:
+  kj::OneOf<kj::Array<kj::byte>, kj::ArrayPtr<const kj::byte>> ownedOrView;
+  kj::Date lastModified;
+
+  bool isWritable() const {
+    // Our file is only writable if it owns the actual data buffer.
+    return ownedOrView.is<kj::Array<kj::byte>>();
+  }
+
+  kj::Array<kj::byte>& writableView() {
+    return JSG_REQUIRE_NONNULL(
+        ownedOrView.tryGet<kj::Array<kj::byte>>(), Error, "Cannot write to a read-only file");
+  }
+
+  kj::ArrayPtr<const kj::byte> readableView() const {
+    KJ_SWITCH_ONEOF(ownedOrView) {
+      KJ_CASE_ONEOF(view, kj::ArrayPtr<const kj::byte>) {
+        return view;
+      }
+      KJ_CASE_ONEOF(owned, kj::Array<kj::byte>) {
+        return owned.asPtr().asConst();
+      }
+    }
+    KJ_UNREACHABLE;
+  }
+};
+
+using ReadableDirectory = DirectoryBase<false>;
+using WritableDirectory = DirectoryBase<true>;
+
+class VirtualFileSystemImpl final: public VirtualFileSystem {
+ public:
+  VirtualFileSystemImpl(kj::Own<Directory>&& root): root(kj::mv(root)) {}
+
+  Directory& getRoot(Lock& js) override {
+    return *root;
+  }
+
+ private:
+  kj::Own<Directory> root;
+};
+}  // namespace
+
+kj::String File::readAllText(Lock& js) const {
+  auto info = stat();
+  KJ_DASSERT(info.type == Node::Type::FILE);
+  JSG_REQUIRE(info.size <= kMaxReadLength, Error, "File too large to read all at once");
+  if (info.size == 0) return kj::String();
+  auto data = kj::heapArray<char>(info.size);
+  auto size = read(js, 0, data.asBytes());
+  JSG_REQUIRE(size == info.size, Error, "failed to read all data");
+  return js.accountedKjString(kj::str(kj::mv(data)));
+}
+
+BufferSource File::readAllBytes(Lock& js) const {
```

Comment:
What about using function overloading the previous function and calling it `readAll` which depends on the return value to determine which function to overload.

Reply:
Much prefer having them separated like this and it's more consistent with every other place where we differentiate reading text vs. reading bytes.